File: Cpp/848_Shifting_Letters.cpp
class Solution {
public:
    string shiftingLetters(string S, vector<int>& shifts) {
        string result="";
        int n = shifts.size();
        vector<int> shift_sum(n, 0);
        shift_sum[n-1] = shifts[n-1] % 26;
        for(int i=n-2; i>=0; --i){
            shift_sum[i] += (shift_sum[i+1] + shifts[i]) % 26;
        }
        
        for(int i=0; i<n; ++i){
            result += 'a' + (S[i]-'a' + shift_sum[i]) % 26;
        }
        
        return result;
    }
};

File: Kotlin/848_Shifting_Letters.kt
class Solution {
    fun shiftingLetters(s: String, shifts: IntArray): String {
        val result = StringBuilder()
        val n = shifts.size
        val shiftSum = LongArray(n)
        
        shiftSum[n - 1] = shifts[n - 1].toLong() % 26
        for(i in n - 2 downTo 0) {
            shiftSum[i] = (shiftSum[i + 1] + shifts[i]) % 26
        }
        
        for(i in shiftSum.indices) {
            result.append((((s[i] - 'a' + shiftSum[i]) % 26) + 'a'.toInt()).toChar())
        }
        
        return result.toString()
    }
}